### PR TITLE
Fix direct run with open_basedir configured

### PR DESCRIPTION
**Pull request by mvorisek**

### CODE REVIEW

File: php-cs-fixer [hunk 1]
```diff
@@ -83,7 +83,7 @@ if (class_exists('Phar')) {
 
 if ($require) {
     // OK, it's not, let give Composer autoloader a try!
-    $possibleFiles = [__DIR__.'/../../autoload.php', __DIR__.'/../autoload.php', __DIR__.'/vendor/autoload.php'];
+    $possibleFiles = [__DIR__.'/vendor/autoload.php', __DIR__.'/../../autoload.php'];
```
Comment (mvorisek) [inline php-cs-fixer:86]: `__DIR__.'/../autoload.php'` should never be needed

File: src/ToolInfo.php [hunk 2]
```diff
@@ -90,7 +90,7 @@ public function isInstalledAsPhar()
     public function isInstalledByComposer()
     {
         if (null === $this->isInstalledByComposer) {
-            $this->isInstalledByComposer = !$this->isInstalledAsPhar() && file_exists($this->getComposerInstalledFile());
+            $this->isInstalledByComposer = !$this->isInstalledAsPhar() && @file_exists($this->getComposerInstalledFile());
```
Comment (mvorisek) [inline src/ToolInfo.php:93]: supress `open_basedir` warning

File: php-cs-fixer (hunk 1 above)
Comment (kubawerlos) [reply php-cs-fixer:86]: > upper directories must be checked later

Why "must"? Previously first check was for most often use case and got hit, so others were not used, after your change first one is path used only in this repository, so everyone using PHP CS Fixer will attempt 1 `file_exists` which will always fail.

> `__DIR__.'/../autoload.php'` should never be needed

See the PR when it was added to see why it is needed: https://github.com/FriendsOfPHP/PHP-CS-Fixer/pull/4490

File: src/ToolInfo.php (hunk 2 above)
Comment (kubawerlos) [reply src/ToolInfo.php:93]: What warning? I'd expect some test added to cover this change.

**Review by kubawerlos — CHANGES_REQUESTED** [review]

File: php-cs-fixer (hunk 1 above)
Comment (mvorisek) [reply php-cs-fixer:86]: > 
> 
> > upper directories must be checked later
> 
> Why "must"?

Easy, if `open_basedir` is set only to this repo, then trying to access any upper directory first will show php warning as the dir is outside.

> Previously first check was for most often use case and got hit, so others were not used, after your change first one is path used only in this repository, so everyone using PHP CS Fixer will attempt 1 `file_exists` which will always fail.

One or two extra file checks, nothing else negative, right?


> > `__DIR__.'/../autoload.php'` should never be needed
> 
> See the PR when it was added to see why it is needed: #4490

I will add this back.

File: src/ToolInfo.php (hunk 2 above)
Comment (mvorisek) [reply src/ToolInfo.php:93]: File access outside `open_basedir` :) I won't write test for this as it will require starting another process.

Comment (kubawerlos) [reply src/ToolInfo.php:93]: So if in a few months someone will remove that `@` all will be fine 😕

Comment (mvorisek) [reply src/ToolInfo.php:93]: Is there any test which tests console with a new php process for each test?

File: php-cs-fixer (hunk 1 above)
Comment (kubawerlos) [reply php-cs-fixer:86]: > if `open_basedir` is set only to this repo

When is that? Was this situation ever reported by anyone? Without any regression check this change makes no sense to me.

> One or two extra file checks, nothing else negative, right?

No one will die, but what is the point if we know that for anyone using PHP CS Fixer as library this check is useless?

Comment (mvorisek) [reply php-cs-fixer:86]: > When is that?

`php -d open_basedir="<current dir>" php-cs-fixer fix` (`open_basedir` is usually set with `php.ini` file, but the effect is the same)

Comment (kubawerlos) [reply php-cs-fixer]: I know how to do it - I meant when it happens in real life? The example you gave means it's either for this repo or someone must have a `php-cs-fixer` file copied to his project's root.

For 1st case there is no `open_basedir` set.
For 2nd case I'd say it's incorrect usage (not sure that even work).

Comment (mvorisek) [reply php-cs-fixer]: 1st case, maby 1st and halt ;-) because of:

> See the PR when it was added to see why it is needed: #4490

**Comment by mvorisek** [comment]
please advise how to test easily or let's merge as is

File: php-cs-fixer [hunk 3]
```diff
@@ -83,7 +83,7 @@ if (class_exists('Phar')) {
 
 if ($require) {
     // OK, it's not, let give Composer autoloader a try!
-    $possibleFiles = [__DIR__.'/../../autoload.php', __DIR__.'/../autoload.php', __DIR__.'/vendor/autoload.php'];
+    $possibleFiles = [__DIR__.'/vendor/autoload.php', __DIR__.'/../autoload.php', __DIR__.'/../../autoload.php'];
```
Comment (keradus) [inline php-cs-fixer:86]: ??

File: src/ToolInfo.php (hunk 2 above)
Comment (keradus) [reply src/ToolInfo.php:93]: why `@` ? isn't warning collected by our `set_error_handler` hook>
no tests, no merge

no, you need to write one from scratch

**Review by keradus — CHANGES_REQUESTED** [review]
vide comments

**Comment by mvorisek** [comment]
@keradus the order is important so so upper directory is not accessed if not needed - otherwise open_basedir emit php warning

File: php-cs-fixer (hunk 3 above)
Comment (mvorisek) [reply php-cs-fixer:86]: https://3v4l.org/DG4MG

Comment (keradus) [reply php-cs-fixer:86]: tests needed

Comment (keradus) [reply php-cs-fixer:86]: maybe also a in-code comment with reasoning

File: src/ToolInfo.php (hunk 2 above)
Comment (SpacePossum) [reply src/ToolInfo.php:93]: The issue is that this only happens if you config `open_basedir` through the PHP `ini` system, which is a pain to do in unit tests. When trying to detect something like we do here what is the harm suppressing the warning?

To reproduce, you configure `open_basedir` to some directory outside of the project install for your CLI and run the tool.

File: php-cs-fixer (hunk 3 above)
Comment (SpacePossum) [reply php-cs-fixer:86]: if we suppress the warning than the order doesn't need to be changed does it @mvorisek ?

**Comment by keradus** [comment]
I leaning toward closing this PR without merging.
looks like PR opener is the only affected person, and I don't want to mess with bootstrap file without having the change covered with automated test, which is still missing

File: src/ToolInfo.php (hunk 2 above)
Comment (SpacePossum) [reply src/ToolInfo.php:93]: @keradus do you know a way to test with different ini values for PHP within a single PHPUnit session?

Comment (keradus) [reply src/ToolInfo.php:93]: i was not checking for that possibility myself

**Comment by keradus** [comment]
I appreciate the effort in this PR, yet I decided to close it due to reasoning shared above. Thank you for understanding.
I'm happy to reconsider the decision when we have more ppl affected or automated test showing issue without this change.

**Comment by mvorisek** [comment]
> 
> 
> please advise how to test easily or let's merge as is

@keradus please help with the test, where can I set `open_basedir` php.ini value?

**Comment by keradus** [comment]
ref https://github.com/FriendsOfPHP/PHP-CS-Fixer/pull/5634#discussion_r681783065

unfortunately, I am not aware of this, I was never in need of that.
I would suggest this as starting point (, but I cannot assist due to lack of experience with `open_basedir`): https://www.php.net/manual/en/ini.core.php#ini.open-basedir

**Comment by mvorisek** [comment]
@keradus ref https://github.com/FriendsOfPHP/PHP-CS-Fixer/pull/5634/files#r679982065

if we add `@` operator before the `file_get_contents`, would that be ok solution /wo test for you?

**Comment by keradus** [comment]
My constant complaint is about lack of test behind proposed change.
Changing code to use silence operator not only make the code worse (damn, we even [escalate](https://github.com/FriendsOfPHP/PHP-CS-Fixer/blob/master/php-cs-fixer#L64) bunch of non-breaking errors to breaking errors!), but also doesn't solve complaint of missing tests.

**Comment by mvorisek** [comment]
@keradus please reopen and I will look for some solution with a test

**Comment by keradus** [comment]
simply raise a new PR when the test is ready

**Comment by SpacePossum** [comment]
You know it cannot be tested https://github.com/FriendsOfPHP/PHP-CS-Fixer/pull/5634#discussion_r681783065 
You cannot change PHP ini values like the one needed for the test @ run time for security reasons, that is the whole point of that ini value. 

Not sure what the issue is here, we want to test if a file exists and PHP triggers an error/warning if that file is in a directory that you are not allowed to access.

**Comment by mvorisek** [comment]
>Not sure what the issue is here, we want to test if a file exists and PHP triggers an error/warning if that file is in a directory that you are not allowed to access.

@SpacePossum the `open_basedir` restrict the FS to a subspace/subpath which is allowed to be accessed. If a file operation is done outside that configuration, an warning is emitted. This clearly reasons https://github.com/FriendsOfPHP/PHP-CS-Fixer/pull/5634/files#diff-375c36cd1ed77ff74277c5cef2521d0832604dd0cc05e56e29255a4b53ff038fR86 to test the more specific path first. This PR is valid, I only do not know how to write a test for it.

**Comment by SpacePossum** [comment]
My question was, what is the issue this get keeping closed? The fix to suppress the `open_dir` seems good to me and would help me out, but somehow there is an impossible requirement that there must be a test.

**Pull request closed without merging**